eventmesh-sdks/eventmesh-sdk-java/src/main/java/org/apache/eventmesh/client/grpc/consumer/SubStreamHandler.java

```diff
@@ -129,12 +130,15 @@ public void run() {
     }
 
     public void close() {
-        if (this.sender != null) {
-            senderOnComplete();
+        synchronized (sender) {
+            if (latch.getCount() == 0) {
+                return;
+            }
```

Comment:
These concurrent programming codes add another layer of `synchronized` outside of `senderOnComplete()` and `CountDownLatch`. This is equivalent to adding another lock outside of the `synchronized` lock, and it wastes the performance advantages of `CountDownLatch` over locks. 
This is because you added `CountDownLatch#getCount()`, which destroys the atomicity of `CountDownLatch` in the original code. The deeper reason for this is the unnecessary circular logic `close() > senderOnComplete() > onCompleted() > close() > ...` that I mentioned earlier.

Reply:
>another lock outside of the `synchronized` lock

`senderOnComplete()` may be called by another caller in the future.

>`CountDownLatch#getCount()` destroys the atomicity of `CountDownLatch` in the original code

The only usage of `latch.countDown()` is included in synchronized block with `latch.getCount() == 0`, so it can be considered atomic. Well maybe we can't prevent somebody call `countDown()` again somewhere else either, so it may be a potential problem.

>optimize this circular logic `close() > senderOnComplete() > onCompleted() > close() > ...`. Analyzing the code reveals that `onCompleted()` does not need to call `close()` again because the key of `close()` method is `latch.countDown()`, so there is no need to go around in a large circle to execute `latch.countDown()`.

You mean replace the `close()` call of `onCompleted()` with a `latch.countDown()`?

I used boolean to prevent latch called by somebody else in https://github.com/apache/eventmesh/pull/4807/commits/4d8694898e20210aced416f71349c4a220257331. Does that sounds good to you?

Reply:
> > another lock outside of the `synchronized` lock (in `senderOnComplete()`)
> 
> `senderOnComplete()` may be called by another caller in the future.

The `synchronized` inside `senderOnComplete()` can ensures safety.

> > `CountDownLatch#getCount()` destroys the atomicity of `CountDownLatch` in the original code
>
> The only usage of `latch.countDown()` is included in synchronized block with `latch.getCount() == 0`, so it can be considered atomic. 

Regarding "`latch.getCount()==0` + `latch.countDown()` can be considered atomic", you can think about it again.

> Well maybe we can't prevent somebody call `countDown()` again somewhere else either, so it may be a potential problem.

When `CountDownLatch` counter reaches 0 and `countDown()` is called again, nothing happens, no exceptions are thrown, and there are no potential problems.

>>optimize this circular logic `close() > senderOnComplete() > onCompleted() > close() > ...`. Analyzing the code reveals that `onCompleted()` does not need to call `close()` again because the key of `close()` method is `latch.countDown()`, so there is no need to go around in a large circle to execute `latch.countDown()`.
>
> You mean replace the `close()` call of `onCompleted()` with a latch.countDown()?

The meaning of this previously mentioned approach is to remove `close()` from `onCompleted()`.

Reply:
May I ask that why a synchronized block with `latch.getCount() == 0` and `latch.countDown()` is not considered atomic?

>this previously mentioned approach is to remove `close()` from `onCompleted()`.

Done in https://github.com/apache/eventmesh/pull/4807/commits/dbabca713d2fe7f1d7b58bf4bc2e371a00a1cc0a.

Reply:
> May I ask that why a synchronized block with `latch.getCount() == 0` and `latch.countDown()` included is not considered atomic?

A synchronized block is naturally atomic. The original sentence is "added `CountDownLatch#getCount()` destroys the atomicity of `CountDownLatch` in the original code", so "`latch.getCount() == 0` + `latch.countDown()`" has to be wrapped with another synchronized.